File: BaseSystem/src/main/java/de/silveryard/basesystem/gui/Shader.java
package de.silveryard.basesystem.gui;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Sebif on 7/13/2017.
 */
public class Shader {
    public static Shader create(String vertexSource, String fragmentSource){
        int program = SDLShader.shaderProgramCreate();
        int vertexShader = SDLShader.createShader(SDLShader.GL_VERTEX_SHADER);
        int fragmentShader = SDLShader.createShader(SDLShader.GL_FRAGMENT_SHADER);

        SDLShader.shaderSetSource(vertexShader, vertexSource);
        SDLShader.shaderSetSource(fragmentShader, fragmentSource);

        int result = SDLShader.shaderCompile(vertexShader);
        if(result < 0)
            throw new RuntimeException("Failed to compile vertex shader!");

        result = SDLShader.shaderCompile(fragmentShader);
        if(result < 0)
            throw new RuntimeException("Failed to compile fragment shader!");

        SDLShader.shaderProgramAttach(program, vertexShader);
        SDLShader.shaderProgramAttach(program, fragmentShader);

        result = SDLShader.shaderProgramLink(program);
        if(result < 0)
            throw new RuntimeException("Failed to link shader program!");

        SDLShader.shaderDelete(vertexShader);
        SDLShader.shaderDelete(fragmentShader);

        return new Shader(program);
    }

    private int program;
    private Map<String, Integer> locations;

    private Shader(int program){
        this.program = program;
        locations = new HashMap<>();
    }

    private int getLocation(String name){
        Integer location = locations.get(name);

        if(location == null){
            location = SDLShader.shaderProgramGetUniformLocation(program, name);
            locations.put(name, location);
        }

        return location;
    }

    public void use(){
        SDLShader.shaderProgramUse(program);
    }

    public void setFloat(String name, float value){
        SDLShader.shaderProgramSetFloat(getLocation(name), value);
    }
    public void setInt(String name, int value){
        SDLShader.shaderProgramSetInt(getLocation(name), value);
    }
    public void setVector2f(String name, float x, float y){
        SDLShader.shaderProgramSetVector2f(getLocation(name), x, y);
    }
    public void setVector3f(String name, float x, float y, float z){
        SDLShader.shaderProgramSetVector3f(getLocation(name), x, y, z);
    }
    public void setVector4f(String name, float x, float y, float z, float w){
        SDLShader.shaderProgramSetVector4f(getLocation(name), x, y, z, w);
    }
}

File: Libraries/System/Native/SDLNative/src/shader.cpp
#define GLEW_STATIC

#include <map>
#include <GL/glew.h>
#include "shader.h"

using namespace std;

SDLNEXPORT int shader_create(int shader_type) {
	GLint type = 0;
	switch (shader_type) {
	case SHADER_TYPE_VERTEX:
		type = GL_VERTEX_SHADER;
		break;
	case SHADER_TYPE_FRAGMENT:
		type = GL_FRAGMENT_SHADER;
		break;
	default:
		return -1;
	}

	GLuint shader = glCreateShader(type);
	return static_cast<int>(shader);
}
SDLNEXPORT void shader_delete(int shader) {
	glDeleteShader(shader);
}
SDLNEXPORT void shader_set_source(int shader, const char* source) {
	GLuint glShader = static_cast<GLuint>(shader);
	const GLchar* shaderSource[] = { source };

	glShaderSource(glShader, 1, shaderSource, NULL);
}
SDLNEXPORT int shader_compile(int shader) {
	GLuint glShader = static_cast<GLuint>(shader);

	glCompileShader(glShader);

	GLint compiled = GL_FALSE;
	glGetShaderiv(glShader, GL_COMPILE_STATUS, &compiled);
	if (compiled != GL_TRUE) {
		printf("Unable to compile vertex shader %d!\n", glShader);
		return -1;
	}
	return 0;
}

SDLNEXPORT int shader_program_create() {
	return static_cast<int>(glCreateProgram());
}
SDLNEXPORT void shader_program_attach(int program, int shader) {
	GLuint glProgram = static_cast<GLuint>(program);
	GLuint glShader = static_cast<GLuint>(shader);

	glAttachShader(glProgram, glShader);
}
SDLNEXPORT int shader_program_link(int program) {
	GLuint glProgram = static_cast<GLuint>(program);

	glLinkProgram(glProgram);

	GLint success = GL_FALSE;
	glGetProgramiv(glProgram, GL_LINK_STATUS, &success);
	if (success != GL_TRUE) {
		printf("Error linking program %d!\n", glProgram);
		return -1;
	}
	return 0;
}

SDLNEXPORT int shader_program_get_uniform_location(int program, const char* name) {
	GLuint glProgram = static_cast<GLuint>(program);

	GLint location = glGetUniformLocation(glProgram, name);
	return static_cast<int>(location);
}
SDLNEXPORT void shader_program_set_float(int location, float value) {
	glUniform1f(location, value);
}
SDLNEXPORT void shader_program_set_int(int location, int value) {
	glUniform1i(location, value);
}
SDLNEXPORT void shader_program_set_vector2f(int location, float x, float y) {
	glUniform2f(location, x, y);
}
SDLNEXPORT void shader_program_set_vector3f(int location, float x, float y, float z) {
	glUniform3f(location, x, y, z);
}
SDLNEXPORT void shader_program_set_vector4f(int location, float x, float y, float z, float w) {
	glUniform4f(location, x, y, z, w);
}

SDLNEXPORT void shader_program_use(int program) {
	glUseProgram(program);
}